### PR TITLE
Dev

### PR DESCRIPTION
**Pull request by tsingcbx99**
add importance weighted adversarial nets. fix bugs for cdan

### CODE REVIEW

File: dalib/adaptation/cdan.py [hunk 1]
```diff
@@ -130,13 +130,14 @@ class RandomizedMultiLinearMap(nn.Module):
 
     def __init__(self, features_dim: int, num_classes: int, output_dim: Optional[int] = 1024):
         super(RandomizedMultiLinearMap, self).__init__()
-        self.Rf = torch.randn(features_dim, output_dim)
-        self.Rg = torch.randn(num_classes, output_dim)
+        device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
```
Comment (JunguangJiang) [inline dalib/adaptation/cdan.py:133]: Why use 'torch.device' here?

File: examples-da/partial/importance_weighted_adversarial_nets.py [hunk 2]
```diff
@@ -0,0 +1,415 @@
...
+        # evaluate on validation set
+        acc1 = validate(val_loader, model, args)
+
+        # remember best acc@1 and save checkpoint
+        if acc1 > best_acc1:
+            best_model = copy.deepcopy(model.state_dict())
+        best_acc1 = max(acc1, best_acc1)
+
+    print("best_acc1 = {:3.1f}".format(best_acc1))
+    model.load_state_dict(best_model)
```
Comment (JunguangJiang) [inline examples-da/partial/importance_weighted_adversarial_nets.py:185]: It's not very fair to load best model as pretrain model since we don't have labels on the target domain.

File: examples-da/partial/importance_weighted_adversarial_nets.py [hunk 3]
```diff
@@ -0,0 +1,415 @@
...
+    classifier_t.head.eval()
+    classifier_t.set_features_only(True)
+
+    domain_adv_D = DomainAdversarialLoss(D).to(device)
+    # define trade off scheduler
+    grl = WarmStartGradientReverseLayer(alpha=1, lo=0, hi=args.trade_off, max_iters=args.epochs * args.iters_per_epoch)
+    domain_adv_D_0 = DomainAdversarialLoss(D_0, grl=grl).to(device)
+    # parameters
+    optimizer_D = SGD(D.get_parameters(), args.lr, momentum=args.momentum, weight_decay=args.weight_decay, nesterov=True)
+    optimizer_D_0 = SGD(classifier_t.get_parameters() + D_0.get_parameters(), args.lr, momentum=args.momentum,
```
Comment (JunguangJiang) [inline examples-da/partial/importance_weighted_adversarial_nets.py:205]: remove '_D_0' might be better since optimizer_D_0  also optimize classifier_t

**Review by JunguangJiang — APPROVED** [review]
